test/unit/client-side-encryption/state_machine.test.ts

```diff
@@ -251,6 +251,73 @@ describe('StateMachine', function () {
       });
     });
 
+    context('when server closed the socket', function () {
+      let server;
+
+      beforeEach(async function () {
+        server = net.createServer(async socket => {
+          socket.end();
+        });
+        server.listen(0);
+        await once(server, 'listening');
+      });
+
+      afterEach(function () {
+        server.close();
+        this.sinon.restore();
+      });
+
+      it('throws a MongoCryptError error with proxyOptions', async function () {
+        const stateMachine = new StateMachine({
+          proxyOptions: {
+            proxyHost: 'localhost',
+            proxyPort: server.address().port
+          }
+        } as any);
+        const request = new MockRequest(Buffer.from('foobar'), 500);
+
+        try {
+          await stateMachine.kmsRequest(request);
+        } catch (err) {
+          expect(err.name).to.equal('MongoCryptError');
+          expect(err.message).to.equal('KMS request failed');
+          return;
+        }
+        expect.fail('missed exception');
+      });
+
+      it('throws a MongoCryptError error with a tls endpoint', async function () {
+        const stateMachine = new StateMachine({
+          tlsOptions: { aws: { tlsCertificateKeyFile: 'test.pem' } }
+        } as any);
+        const request = new MockRequest(Buffer.from('foobar'), -1);
+        const buffer = Buffer.from('foobar');
+
+        this.sinon.stub(fs, 'readFile').callsFake(fileName => {
+          expect(fileName).to.equal('test.pem');
+          return Promise.resolve(buffer);
+        });
+        this.sinon.stub(tls, 'connect').callsFake((options, callback) => {
+          this.fakeSocket = new MockSocket(callback);
```

Comment:
We are returning a MockSocket here instead of one connected to the mock server, what is the mock server providing?

Reply:
I refactored this test so we actually close the socket on server and check the failure message.